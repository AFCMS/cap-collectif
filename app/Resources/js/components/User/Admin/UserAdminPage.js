// @flow
import * as React from 'react';
import { type ReadyState, QueryRenderer, graphql } from 'react-relay';
import { connect } from 'react-redux';
import { isDirty } from 'redux-form';
import environment, { graphqlError } from '../../../createRelayEnvironment';
import UserAdminPageTabs from './UserAdminPageTabs';
import Loader from '../../Ui/FeedbacksIndicators/Loader';
import type { State } from '../../../types';
import type {
  UserAdminPageQueryResponse,
  UserAdminPageQueryVariables,
} from '~relay/UserAdminPageQuery.graphql';

export type Props = { userId: string, dirty: boolean };

const onUnload = e => {
  e.returnValue = true;
};

const component = ({ error, props }: { props?: ?UserAdminPageQueryResponse } & ReadyState) => {
  if (error) {
    console.log(error); // eslint-disable-line no-console
    return graphqlError;
  }
  if (props) {
    // eslint-disable-next-line
    if (props.user !== null) {
      return <UserAdminPageTabs {...props} />;
    }
    return graphqlError;
  }
  return <Loader />;
};

export class UserAdminPage extends React.Component<Props> {
  componentDidUpdate(prevProps: Props) {
    if (prevProps.dirty === false && this.props.dirty === true) {
      window.addEventListener('beforeunload', onUnload);
    }

    if (this.props.dirty === false) {
      window.removeEventListener('beforeunload', onUnload);
    }
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', onUnload);
  }

  render() {
    return (
      <div className="admin_proposal_form">
        <QueryRenderer
          environment={environment}
          query={graphql`
            query UserAdminPageQuery($id: ID!) {
              user: node(id: $id) {
                ...UserAdminPageTabs_user
              }
            }
          `}
          variables={
            ({
              id: this.props.userId,
            }: UserAdminPageQueryVariables)
          }
          render={component}
        />
      </div>
    );
  }
}

const mapStateToProps = (state: State) => ({
  dirty:
    isDirty('user-admin-edit')(state) ||
    isDirty('user-admin-selections')(state) ||
    isDirty('user-admin-evaluation')(state) ||
    isDirty('user-admin-status')(state),
});

export default connect(mapStateToProps)(UserAdminPage);
