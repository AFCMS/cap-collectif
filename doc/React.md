### How to write localizable components

Just import the appropriate [component](https://github.com/yahoo/react-intl/wiki#the-react-intl-module) from `react-intl`

- For localizable text use
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage).

- For date and time:
[`<FormattedDate>`](https://github.com/yahoo/react-intl/wiki/Components#formatteddate)
[`<FormattedTime>`](https://github.com/yahoo/react-intl/wiki/Components#formattedtime)
[`<FormattedRelative>`](https://github.com/yahoo/react-intl/wiki/Components#formattedrelative)

- For numbers and currencies:
[`<FormattedNumber>`](https://github.com/yahoo/react-intl/wiki/Components#formattednumber)
[`<FormattedPlural>`](https://github.com/yahoo/react-intl/wiki/Components#formattedplural)

- If possible, do not use `<FormattedHTMLMessage>`, see how to use *Rich Text Formatting* with
[`<FormattedMessage>`](https://github.com/yahoo/react-intl/wiki/Components#formattedmessage)

- When you need an imperative formatting API, use the [`injectIntl`](https://github.com/yahoo/react-intl/wiki/API#injectintl) High-Order Component.
