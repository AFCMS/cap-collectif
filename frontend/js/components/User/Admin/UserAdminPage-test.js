/* eslint-env jest */
/* @flow */
import React from 'react';
import { shallow } from 'enzyme';
import { UserAdminPage } from './UserAdminPage';

describe('<UserAdminPage/>', () => {
  it('should render', () => {
    const wrapper = shallow(<UserAdminPage userId="user516" dirty={false} />);
    wrapper.setState({
      showModal: true,
    });
    expect(wrapper).toMatchSnapshot();
  });
});
