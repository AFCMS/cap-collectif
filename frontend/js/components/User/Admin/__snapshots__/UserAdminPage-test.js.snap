// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAdminPage/> should render 1`] = `
<div
  className="admin_proposal_form"
>
  <ReactRelayQueryRenderer
    environment={"RelayModernEnvironment()"}
    query={[Function]}
    render={[Function]}
    variables={
      Object {
        "id": "userMaxime",
      }
    }
  />
</div>
`;
